### ISSUE TITLE
0.12.7 -> 0.12.8 now strange behavior with dynamic component

### ISSUE THREAD

**Issue by fullfs** [yyx990803/vue#1098]
Hello!
Can't say for sure what's going on, but I've got the problem.
In 0.12.7 everything is ok, but in 0.12.8 my app started to lag a lot and now I've got warnings like 
```You may have an infinite update loop for the watcher with expression "offersList"```

The code I've got most warnings for looks like this:

```
<div
    v-if="isOffersPresent"
    v-component="{{offersView}}"
    offers-list="{{offers.result}}"
></div>
```

So it's a dynamic component with dynamic props and ```replace: false```

Any hints?

**Comment by fullfs** [comment]
Got it on fiddle!
http://jsfiddle.net/wvbqbyLp/23/

For some strange reason the problem is dependent on number of items to be repeated inside ```item``` and on number of props ```item``` has got inside its ```data``` function.

O_o

**Comment by yyx990803** [comment]
If you comment out these lines: https://github.com/yyx990803/vue/blob/dev/src/batcher.js#L73-L79
Does the app still lag?

**Comment by fullfs** [comment]
@yyx990803 Aye,still lag. Warning is different now: "Duplicate objects are not supported in v-repeat when using components or transitions"

**Comment by fullfs** [comment]
@yyx990803 And another thing: The app now freezes in infinite loop with those warnings

**Comment by yyx990803** [comment]
Can you try this branch: https://github.com/yyx990803/vue/tree/fix-batcher-infinite-check

It should have fixed the infinite check, just want to make sure it also fixed the performance problem.

**Comment by fullfs** [comment]
@yyx990803 Didn't help :( Infinite check is still present.
Probably should investigate more. The problem occurs only at one page. We're also using a few hacks to work with ES6, maybe those hacks in the way.

**Comment by fullfs** [comment]
@yyx990803 Ah, set ```Vue.config.async``` to false. This is where the fun begins.

**Comment by yyx990803** [comment]
Hmm, it fixed your fiddle though. Are you using the file in dist? I didn't
build in that branch.

**Comment by fullfs** [comment]
@yyx990803 I did build it with ```grunt build```, yeah.
I've tried new version in separated document on another computer, can't get it right new.
I've also tried to reproduce the example right now, but can't even get Vue v0.12.8 to work in async=false mode. Getting ```Cannot read property 'appendChild' of null``` error

**Comment by fullfs** [comment]
http://jsfiddle.net/wvbqbyLp/24/
Added ```Vue.config.async = false```
Getting ```Cannot read property 'appendChild' of null``` error too. Looks like on the current computer the code behaves different. Before I did get inifinite loop with the same code.

**Comment by yyx990803** [comment]
`async: false` probably doesn't help here.

The fiddle is now working properly in the fix branch, if your app is still having an infinite loop, then it probably does have an infinite loop... you will need to make a new fiddle showing infinite loop problem with the fix branch.

Some low-level details:

In your fiddle, your `item` component has a lot of default data fields; these fields are being merged into the objects in the `offers` array. Every field causes an `$add` call. In 0.12.7, when you `$add` an object inside the array, the watcher watching that array is not triggered. In 0.12.8 it does, so every `$add` call triggers the `v-repeat` to be updated again - which is why the infinite check has something to do with the number of elements in the array. This is also why there is a performance issue.

The fix branch makes sure a watcher is only updated once every async update cycle, and only counts infinite loops if a watcher pushes itself into the queue again.

**Comment by fullfs** [comment]
@yyx990803 thanks for detailed explanation! It makes some things clear to me.
But still, I have some points.
1. Could you please give me working fiddle with my code example and ```Vue.config.async = false```? I have some thoughts about the problem not always being visible.

2. I want to make accent to following points: I've tested 2 versions of Vue.js: ```0.12.8 regular``` and ```fix-batcher-infinite-check```. There were also 2 versions of code: with async ```false``` and ```true```. So I've made 4 different tests. So, lets review the results:
```0.12.8 regular``` & ```async = true```: bunch of ```You may have an infinite update loop for the watcher with expression "offersList"``` errors;
```0.12.8 regular``` & ```async = false```: infinite loop or ```Cannot read property 'appendChild' of null``` error (depends on computer)

```fix-batcher-infinite-check``` & ```async = false```: Everything is fine! Yay!

```fix-batcher-infinite-check``` & ```async = true```: infinite loop or ```Cannot read property 'appendChild' of null``` error (depends on computer)


So either I've failed in building ```fix-batcher-infinite-check``` properly or the problems just don't occur on your system.

**Comment by yyx990803** [comment]
I can see the point: yes, when setting `async = false` I get an error with the fix branch too.
Are you using `async = false` in your actual app? Honestly I don't recommend doing so exactly because of this type of situation.

**Comment by fullfs** [comment]
@yyx990803 Well yes I do. Actually I didn't intended to. I've used ```Vue.config.async = true``` from the beginning. The thing is so funny! TIll version 0.12 Vue.js had a bug (probably) that components used inside wait-for component would ignore ```Vue.config.async = true``` and work synchronously. After trying to move to v 0.12 I've found my app not working because in the new version ```Vue.config.async = true``` finnaly began to work, while the whole app now is designed to work depending on synchronous Vue.js behavior.
